watch: fix watch args not being properly filtered

dario-piotrowicz opened the pull request:
Currently the [logic for filtering out watch related flags](https://github.com/nodejs/node/blob/25842c5e35efb45df169e591c775a3c4f853556d/lib/internal/main/watch_mode.js#L42-L55) before passing them to the watch target script is flawed, as it includes any flag (starting with `-`) that follows a flag starting with `--watch` ([code](https://github.com/nodejs/node/blob/25842c5e35efb45df169e591c775a3c4f853556d/lib/internal/main/watch_mode.js#L49-L51)), this allows some watch flags to make it pass the filtering resulting in the generation of multiple watch processes which is both wasteful and can cause duplicate terminal logs (alongside other issues if the target script is side-effectful).

For example if I run
```sh
node --watch --watch index.js
```
the current implementation [will spawn](https://github.com/nodejs/node/blob/25842c5e35efb45df169e591c775a3c4f853556d/lib/internal/main/watch_mode.js#L69) a process equivalent to:
```sh
node --watch index.js
```
which will then finally spawn the correct:
```sh
node index.js
```

Here I'm addressing the incorrect filtering so that all the watch flags are correctly filtered out (making the extra process spawning disappear)

___

Fixes https://github.com/nodejs/node/issues/57124

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

dario-piotrowicz commented on test/sequential/test-watch-mode.mjs:
```diff
@@ -791,4 +791,54 @@ process.on('message', (message) => {
     ]);
   });
+
+  it('when multiple `--watch` flags are provided should run as if only one was', async () => {
```
Note: maybe node should error if multiple `--watch` flags are provided?

I did not go with such a solution because that would be a breaking change, just ignoring extra `--watch` flags is a non breaking change and it solves the issue at hand

If someone believes that multiple `--watch` flags should cause an error I would propose to first land this change right now as a patch then as a followup I can also add the stricter validation which will be included as a breaking change in the next major 🙂

dario-piotrowicz replied:
And/or maybe, the erroring when multiple `--watch` flags are provided is part of the bigger issues of node being quite permissive with its cli flags: https://github.com/nodejs/node/issues/57864 and could be addressed as part of that 🤔

joyeecheung commented on test/sequential/test-watch-mode.mjs:
```diff
@@ -791,4 +791,54 @@ process.on('message', (message) => {
     ]);
   });
+
+  it('when multiple `--watch` flags are provided should run as if only one was', async () => {
```
I am not sure if it's this PR, but `test-watch-mode` has been extremely flaky in both Jenkins and GitHub actions in the past two days e.g. https://ci.nodejs.org/job/node-test-commit/79645/ https://ci.nodejs.org/job/node-test-commit/79644/ and many others. Please refrain from appending new test cases in existing test files as https://github.com/nodejs/node/blob/main/doc/contributing/writing-tests.md suggests - especially this file because it's already being ignored by Jenkins so you are not going to get any proper coverage by appending test cases  here.

dario-piotrowicz replied:
I see, thank you very much and sorry for the trouble, I will keep this in mind and create tests in new files whenever possible 🙇

joyeecheung commented:
Sending a revert since it should not have landed https://github.com/nodejs/node/pull/58190

dario-piotrowicz commented:
> Actually, this PR landed despite failing many platforms in the test file it appends to: https://ci.nodejs.org/job/node-test-pull-request/66587/

Note: I've addressed those issues in https://github.com/nodejs/node/pull/58183, I wonder if there were actually other issues as well :thinking:
